MPR#7642: on ARMv4 and ARMv5, wrong register allocation for integer multiply

xavierleroy opened the pull request:
Consider the ARM integer multiply and multiply-accumulate instructions:
```
         mul rd, rm, rs
         mla rd, rm, rs, ra
```
Before ARMv6, the destination register rd must be different from the
first argument register rm.  Otherwise, undefined behavior could occur,
and actually the assembler refuses the instructions.

In asmcomp/arm/selection.ml, this constraint rd <> rm is cleverly
expressed by pretending that the first argument to the multiply
is also an additional result.  Hence, for `X <- Y * Z`, selection produces
```
        (X, Y) <- Y * Z
```
Then, register allocation knows not to put pseudoregisters X and Y in
the same register.

The problem, exhibited in [MPR#7642](https://caml.inria.fr/mantis/view.php?id=7642), is that this invariant "second result
of multiply is first argument" is not preserved by the reloading phase.
Assume that on the first round of register allocation, Y was spilled to
stack.  Then, reloading takes place and produces two different
pseudo-registers Y1 and Y2 for uses of Y before and after the multiply:
```
        // use Y1
        (X, Y2) <- Y1 * Z
        // use Y2
```
The second round of register allocation gladly puts Y1 and Y2 in different
registers or stack locations.  Uses of Y2 after the multiply therefore
use the wrong value of Y, because Y2 was never initialized to begin with!

This commit fixes the issue by adding a special reloading rule for multiply
and multiply-accumulate instructions that have 2 results.  The special
rule ensures that the first argument and the second result are reloaded
to the same pseudoregister.  (E.g. that Y1=Y2 in the example above.)
This way, the invariant "second result of multiply is first argument"
is preserved all the way to assembly code emission.

In the original report (MPR#7642) the problem is tied to

1. generation of position-independent code (PIC)
2. the "armel" (soft-FP ABI) Debian port as opposed to the "armhf" (hard-FP EABI) port.

Both factors are mostly unrelated.  PIC changes register availability
and therefore produces a different register allocation than non-PIC.
The PIC register allocation happens to trigger the bug on function
Cmmgen.udivmod from the OCaml sources.

Concerning "armel" vs "armhf", "armhf" is available for ARMv6 and above,
hence will never run into the bug.  "armel", on the other hand, still
defaults to ARMv4T.  Configuring OCaml for, say, armv7l-unknown-linux-gnueabi
(ARMv7 + soft-FP EABI) would avoid the bug as well.  It shows up in
Debian's armel only because they configure for arm-unknown-linux-gnueabi,
which is their ARMv4T base line.

xavierleroy committed:
MPR#7642: on ARMv4 and ARMv5, wrong register allocation for integer multiply

Consider the ARM integer multiply and multiply-accumulate:
         mul rd, rm, rs
         mla rd, rm, rs, ra

Before ARMv6, the destination register rm must be different from the
first argument register rm.  Otherwise, undefined behavior could occur,
and actually the assembler refuses the instructions.

In asmcomp/arm/selection.ml, this constraint rd <> rm is cleverly
expressed by pretending that the first argument to the multiply
is also an additional result.  Hence, for X <- Y * Z, selection produces
        (X, Y) <- Y * Z
Then, register allocation knows not to put pseudoregisters X and Y in
the same register.

The problem, exhibited in MPR#7642, is that this invariant "second result
of multiply is first argument" is not preserved by the reloading phase.
Assume that on the first round of register allocation, Y was spilled to
stack.  Then, reloading takes place and produces two different
pseudo-registers Y1 and Y2 for uses of Y before and after the multiply:
        // use Y1
        (X, Y2) <- Y1 * Z
        // use Y2
The second round of register allocation gladly puts Y1 and Y2 in different
registers or stack locations.  Uses of Y2 after the multiply therefore
use the wrong value of Y, because Y2 was never initialized to begin with!

This commit fixes the issue by adding a special reloading rule for multiply
and multiply-accumulate instructions that have 2 results.  The special
rule ensures that the first argument and the second result are reloaded
to the same pseudoregister.  (E.g. that Y1=Y2 in the example above.)
This way, the invariant "second result of multiply is first argument"
is preserved all the way to assembly code emission.

In the original report (MPR#7642) the problem is tied to
1- generation of position-independent code (PIC)
2- the "armel" (soft-FP ABI) Debian port
   as opposed to the "armhf" (hard-FP EABI) port.
Both factors are mostly unrelated.  PIC changes register availability
and therefore produces a different register allocation than non-PIC.
The PIC register allocation happens to trigger the bug on function
Cmmgen.udivmod from the OCaml sources.

Concerning "armel" vs "armhf", "armhf" is available for ARMv6 and above,
hence will never run into the bug.  "armel", on the other hand, still
defaults to ARMv4T.  Configuring OCaml for, say, armv7l-unknown-linux-gnueabi
(ARMv7 + soft-FP EABI) would avoid the bug as well.  It shows up in
Debian's armel only because they configure for arm-unknown-linux-gnueabi,
which is their ARMv4T base line.

xavierleroy commented:
I learnt from the masters of the novella-sized commit message :-)

xavierleroy commented:
By the way and before you ask: I'm not adding any tests, on purpose.  I actually have a small repro case, but it triggers only on ARMv4 or ARMv5 / armel and our CI machines are ARMv7 / armhf.

gasche commented:
@mshinwell, would you by chance be able to review this?

mshinwell commented:
Thanks for looking at this whilst I was on vacation.  I am currently reviewing the patch.

gasche reviewed (approved):
I could check that the implementation corresponds to the explanation, and given that this only applies to barely-supported ARM versions (and fixes a critical bug for these versions) I think that it is reasonable to merge it in 4.06.

mshinwell commented:
Please hold on, I am still thinking about this one.

mshinwell commented:
@xavierleroy The proposed change looks correct to me.  However I don't understand the following.  In `arm/selection.ml` there are also register constraints in the following two cases, one of which applies to armhf platforms:
```
  (* Soft-float Iabsf and Inegf: arg.(0) and res.(0) must be the same *)
  | Iabsf | Inegf when !fpu = Soft ->
      ([|res.(0); arg.(1)|], res)
  (* VFPv{2,3} Imuladdf...Inegmulsubf: arg.(0) and res.(0) must be the same *)
  | Ispecific(Imuladdf | Inegmuladdf | Imulsubf | Inegmulsubf) ->
      let arg' = Array.copy arg in
      arg'.(0) <- res.(0);
      (arg', res)
```
Is it not the case that the reload pass is able to break these constraints if the register concerned gets spilled during register allocation?  There are no special cases in `arm/reload.ml` for these at the moment.  (I'm not sure of what the impact of this would be---presumably the assembler rejects the instructions to avoid undefined behaviour.)

xavierleroy commented:
@mshinwell : you're right, the selection rules you quote above need to be reflected as special reload rules.  Basically, we have "two-address" instructions and the "two-address" constraint that result = argument1 must be preserved.  I'll work on this tomorrow.

xavierleroy committed:
MPR#7642 continued: other possibly wrong register allocations on ARM

asmcomp/arm/reload.ml is modified to preserve "two-address" constraints of the form "this argument and this result must be the same register", constraints that are initially enforced by selection.ml and later assumed by emit.mlp.

MPR#7642 is an instance of this problem involving integer multiplication on ARMv4 and ARMv5.  It is addressed by commit b525cb6.  This commit addresses two other instances of the problem (reported by Mark Shinwell in GPR#1411):  FP multiply-and-add on VFD, and FP neg/abs in soft FP emulation.

xavierleroy commented:
I added the appropriate reloading rules for the other "two address" situations spotted by @mshinwell.

mshinwell commented:
I will review the new code either later today or tomorrow.

damiendoligez commented on Changes:
```diff
@@ -336,8 +336,12 @@ Release branch for 4.06:
+  multiply on ARMv4 and ARMv5; possible wrong register allocation for
+  floating-point multiply and add on VFD and for floating-point
+  negation and absolute value on soft FP emulation.
+  (Xavier Leroy, report by Stéphane Glondu and Mantis user "infinity0",
```
You should use single-quotes instead of double-quotes.
cc @gasche (there is another occurrence in `Changes`)

gasche replied:
I will make a pass to translate all known usernames (only users that wish to remain pseudonymous are kept as pseudonyms) later, so it does not matter -- I don't want to draw Xavier's ire by being too picky about his Changes entries. According to their github page, "infinity0" goes by the name Ximin Luo.

xavierleroy replied:
Thank you for your leniency concerning name quoting, much appreciated.

mshinwell commented on asmcomp/arm/emit.mlp:
```diff
@@ -609,6 +609,7 @@ let emit_instr i =
         let instr = name_for_int_operation op in
         `	{emit_string instr}	{emit_reg i.res.(0)}, {emit_reg i.arg.(0)}, #{emit_int n}\n`; 1
     | Lop(Iabsf | Inegf as op) when !fpu = Soft ->
+        assert (i.res.(0).loc = i.arg.(0).loc);
```
I suggest we add a similar assertion (but an inequality) for `Iintop Imul` and `Ispecific Imuladd` when `!arch < ARMv6`.

xavierleroy replied:
That's not necessary because the assembler will give an error if the architectural constraint is not met. (Like it did in the original MPR.)  And the error message produced by the assembler is more legible than "assertion failed".  

For the other two cases (those found by @mshinwell) the assertion is much more valuable because the error cannot be seen on the asm output.

mshinwell commented on asmcomp/arm/reload.ml:
```diff
@@ -33,12 +34,23 @@ method! reload_operation op arg res =
       if Array.length res' >= 2 then res'.(1) <- arg'.(0);
       argres'
+  | Ispecific(Imuladdf | Inegmuladdf | Imulsubf | Inegmulsubf) ->
+      (* VFD float multiply-add instructions are "two-address" in the
```
VFP not VFD

xavierleroy replied:
Fixing right now.

mshinwell reviewed (approved):
OK to merge after comments addressed.

mshinwell commented:
@damiendoligez @gasche This should go onto 4.06.

xavierleroy committed:
MPR#7642 continued: fixing typos

It's VFP, not VFD.

xavierleroy commented:
Cherry-picked in [4.06 842e995]
